File: word-search-ii/word-search-ii.java
class TrieNode {
    public boolean isWord;
    public TrieNode[] children;

    public TrieNode() {
        isWord = false;
        children = new TrieNode[26];
    }
}

class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void add(String word) {
        TrieNode temp = root;
        for (char c : word.toCharArray()) {
            if (temp.children[c - 'a'] == null) {
                temp.children[c - 'a'] = new TrieNode();
            }
            temp = temp.children[c - 'a'];
        }
        temp.isWord = true;
    }

    public boolean searchWord(String word) {
        TrieNode temp = search(word);
        return temp != null && temp.isWord;
    }

    public boolean searchPrefix(String word) {
        TrieNode temp = search(word);
        return temp != null;
    }

    public TrieNode search(String word) {
        TrieNode temp = root;
        for (char c : word.toCharArray()) {
            if (temp.children[c - 'a'] == null) {
                return null;
            }
            temp = temp.children[c - 'a'];
        }
        return temp;
    }
}

class Solution {
    List<String> ans = new ArrayList<String>();
    int[][] directions = {{0,1 }, {1, 0}, {-1, 0}, {0, -1}};
    public List<String> findWords(char[][] board, String[] words) {
        if (board.length == 0 || board[0].length == 0 || words.length == 0) return ans;
        Trie trie = new Trie();

        for (String word : words) {
            trie.add(word);
        }

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                boolean[][] visited = new boolean[board.length][board[0].length];
                dfs(visited, trie, board, i, j, "");
            }
        }
        return ans;
    }

    public void dfs(boolean[][] visited, Trie trie, char[][] board, int i, int j, String soFar) {
        if (visited[i][j]) return;
        visited[i][j] = true;
        soFar = soFar + board[i][j];
        if (!trie.searchPrefix(soFar)) {
            visited[i][j] = false;
            return;
        }
        if (trie.searchWord(soFar)) {
            ans.add(soFar);
            TrieNode node = trie.search(soFar);
            node.isWord = false;
        }
        for (int[] dir : directions) {
            int nexti = i + dir[0];
            int nextj = j + dir[1];
            if (nexti >= 0. && nexti < board.length && nextj >= 0 && nextj < board[0].length) {
                dfs(visited, trie, board, nexti, nextj, soFar);
            }
        }
        visited[i][j] = false;
    }
}

File: 212-word-search-ii/word-search-ii.py
class TrieNode:
    def __init__(self, parent=None):
        self.parent = parent
        self.children = {}
        self.isWord = False
class Trie:
    def __init__(self):
        self.root = TrieNode()
        self.pointer = self.root
    def insert(self, word: str):
        curr = self.root
        for char in word:
            if char in curr.children:
                curr = curr.children[char]
            else:
                curr.children[char] = TrieNode(curr)
                curr = curr.children[char]
        curr.isWord = True
    # def startsWith(self, prefix):
    #     curr = self.root
    #     for char in prefix:
    #         if char not in curr.children:
    #             return False
    #         curr = curr.children[char]
    #     return True
    def hasNextChar(self, char):
        if char in self.pointer.children:
            self.pointer = self.pointer.children[char]
            self.last = self.pointer
            return True
        return False
    def resetPointer(self):
        self.pointer = self.pointer.parent
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        m = len(board)
        n = len(board[0])
        directions = ((0, 1), (1, 0), (-1, 0), (0, -1))
        res = set()
        visited = set()
        word_set = set(words)
        trie = Trie()
        for word in words:
            trie.insert(word)
        def dfs(row, col, prefix):
            if not trie.hasNextChar(board[row][col]):
                return
            if prefix in word_set:
                res.add(prefix)
            visited.add((row, col))
            for di in directions:
                next_row = row + di[0]
                next_col = col + di[1]
                if 0 <= next_row < m and 0 <= next_col < n and (next_row, next_col) not in visited:
                    dfs(next_row, next_col, prefix + board[next_row][next_col])
            visited.remove((row, col))
            trie.resetPointer()
        for i in range(m):
            for j in range(n):
                dfs(i, j, board[i][j])
                #visited.clear()
        return list(res)